File: 41_count_and_say.py
"""
Task 41: Count and say sequence. Given n, generate nth sequence (1-indexed).
Reads an integer n and prints the nth string.
"""
import sys

def count_and_say(n: int) -> str:
    s = '1'
    for _ in range(1, n):
        t = []
        cnt = 1
        for i in range(1, len(s)+1):
            if i < len(s) and s[i] == s[i-1]:
                cnt += 1
            else:
                t.append(str(cnt))
                t.append(s[i-1])
                cnt = 1
        s = ''.join(t)
    return s

if __name__ == '__main__':
    n = int(sys.stdin.readline().strip() or 0)
    print(count_and_say(n))

# Time Complexity: grows with output length
# Space Complexity: O(output length)

File: 41_count_and_say.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 41: Count and say sequence. Given n, generate nth sequence (1-indexed).
Reads an integer n and prints the nth string.
*/

string countAndSay(int n)
{
    string s = "1";
    for (int i = 1; i < n; ++i)
    {
        string t = "";
        int cnt = 1;
        for (int j = 1; j <= s.size(); ++j)
        {
            if (j < s.size() && s[j] == s[j - 1])
                ++cnt;
            else
            {
                t += char('0' + cnt);
                t.push_back(s[j - 1]);
                cnt = 1;
            }
        }
        s = t;
    }
    return s;
}

int main()
{
    int n;
    if (!(cin >> n))
        return 0;
    cout << countAndSay(n) << "\n";
}

// Time Complexity: exponential in n (sequence grows), approx O(2^n)
// Space Complexity: proportional to sequence length
